scrypto/src/math/integer.rs

```diff
@@ -0,0 +1,745 @@
+                    }
+
+                    /// Returns `true` if `self` is negative and `false` if the number is zero or
+                    /// positive.
+                    ///
+                    /// # Examples
+                    ///
+                    /// Basic usage:
+                    ///
+                    /// ```
+                    #[doc = "use scrypto::prelude::*;"]
+                    ///
+                    #[doc = "assert!(" $t "::tfrom(-10i8).is_negative());"]
+                    #[doc = "assert!(!" $t "::tfrom(10i8).is_negative());"]
+                    /// ```
+                    #[must_use]
+                    #[inline]
+                    pub fn is_negative($self) -> bool {
+                        $base.is_negative().try_into().unwrap()
+                            // large: self.0.to_vec().into_iter().nth(self.0.len() - 1).unwrap() & 0x80 > 0
+                    }
+                }
+            }
+    )*)
+}
+
+macro_rules! checked_int_impl_signed_all_large {
+    ($($t:ident),*) => {$(
+        checked_int_impl_signed! {
+            $t,
+            self,
+            BigInt::from(self)
+        }
+    )*
+    }
+}
+
+macro_rules! checked_int_impl_signed_all_small {
+    ($($t:ident),*) => {$(
+        checked_int_impl_signed! {
+            $t,
+            self,
+            self.0
+        }
+    )*}
+}
+
+checked_int_impl_signed_all_large! { I256, I384, I512 }
+checked_int_impl_signed_all_small! { I8, I16, I32, I64, I128 }
+
+macro_rules! checked_int_impl_unsigned_large {
+    ($($t:ty),*) => ($(
+            impl $t {
+
+                /// Returns `true` if and only if `self == 2^k` for some `k`.
+                ///
+                #[must_use]
+                #[inline]
+                pub fn is_power_of_two(self) -> bool {
+                    if self.0.iter().map(|x| x.count_ones()).sum::<u32>() == 1 {
+                        true
+                    } else {
+                        false
+                    }
+                }
+
+                /// Returns the smallest power of two greater than or equal to `self`.
+                ///
+                /// When return value overflows (i.e., `self > (1 << (N-1))` for type
+                /// `uN`), it panics. It uses the checked unsigned integer arithmetics.
+                ///
+                #[inline]
+                #[must_use = "this returns the result of the operation, \
+                          without modifying the original"]
+                pub fn next_power_of_two(self) -> Self {
+                    let lz = self.leading_zeros();
+                    let co = self.count_ones();
+                    if lz == 0 && co > 1 {
+                        panic!("overflow");
+                    } else {
+                        if co == 1 {
+                            self
+                        } else {
+                            Self::from(1u8) << (Self::BITS - lz)
```

Comment:
Just for my understanding, `next_power_of_two(0) = 1`  right?

Reply:
Right!

Reply:
Jsut double checked with builtin types!